punchcut/src/source/glyf/deltas.rs

```diff
@@ -0,0 +1,289 @@
+use read_fonts::{
+    tables::glyf::{PointFlags, PointMarker},
+    tables::gvar::{GlyphDelta, Gvar},
+    types::{F26Dot6, F2Dot14, Fixed, GlyphId, Point},
+    ReadError,
+};
+
+/// Compute a set of deltas for the component offsets of a composite glyph.
+///
+/// Interpolation is meaningless for component offsets so this is a
+/// specialized function that skips the expensive bits.
+pub fn composite_glyph(
+    gvar: &Gvar,
+    glyph_id: GlyphId,
+    coords: &[F2Dot14],
+    deltas: &mut [Point<Fixed>],
+) -> Result<(), ReadError> {
+    for delta in deltas.iter_mut() {
+        *delta = Default::default();
+    }
+    let Ok(var_data) = gvar.glyph_variation_data(glyph_id) else {
+        // Empty variation data for a glyph is not an error.
+        return Ok(());
+    };
+    for tuple in var_data.tuples() {
+        let Some(scalar) = tuple.compute_scalar(coords) else {
+            continue;
+        };
+        if tuple.all_points() {
+            for (delta, tuple_delta) in deltas.iter_mut().zip(tuple.deltas()) {
+                *delta += tuple_delta.apply_scalar(scalar);
+            }
+        } else {
+            for tuple_delta in tuple.deltas() {
+                let ix = tuple_delta.position as usize;
+                if let Some(delta) = deltas.get_mut(ix) {
+                    *delta += tuple_delta.apply_scalar(scalar);
+                }
+            }
+        }
+    }
+    Ok(())
+}
+
+pub struct SimpleGlyph<'a> {
+    pub points: &'a [Point<i32>],
+    pub flags: &'a mut [PointFlags],
+    pub contours: &'a [u16],
+}
+
+/// Compute a set of deltas for the points in a simple glyph.
+///
+/// This function will use interpolation to infer missing deltas for tuples
+/// that contain sparse sets. The `working_points` buffer is temporary storage
+/// used for this and the length must be >= glyph.points.len().
+pub fn simple_glyph(
+    gvar: &Gvar,
+    glyph_id: GlyphId,
+    coords: &[F2Dot14],
+    has_var_lsb: bool,
+    glyph: SimpleGlyph,
+    working_points: &mut [Point<Fixed>],
+    deltas: &mut [Point<Fixed>],
+) -> Result<(), ReadError> {
+    for delta in deltas.iter_mut() {
+        *delta = Default::default();
+    }
+    let Ok(var_data) = gvar.glyph_variation_data(glyph_id) else {
+        // Empty variation data for a glyph is not an error.
+        return Ok(());
+    };
+    let SimpleGlyph {
+        points,
+        flags,
+        contours,
+    } = glyph;
+    // Include the first phantom point if the font is missing variable metrics
+    // for left side bearings. The adjustment made here may affect the final
+    // shift of the outline.
+    let actual_len = if has_var_lsb {
+        points.len() - 4
+    } else {
+        points.len() - 3
+    };
+    for tuple in var_data.tuples() {
+        let Some(scalar) = tuple.compute_scalar(coords) else {
+            continue;
+        };
```

Comment:
You duplicate the notion of looping over all the tuples that are applicatable. At a glance I believe you could switch from:

```rust
    for tuple in var_data.tuples() {
        let Some(scalar) = tuple.compute_scalar(coords) else {
            continue;
        };
```

To something like `for (tuple, scalar) in var_data.active_tuples_at(coord)` in both simple and composite.

Reply:
pinot actually has a method that does this but that didn't make it over to the implementation here in the first cut. Done!